File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys
input = sys.stdin.readline

n = int(input().rstrip())
wine = [0] * 10001
for i in range(n):
    wine[i] = int(input().rstrip())

dp = [0] * 10001
dp[0] = wine[0]
dp[1] = dp[0] + wine[1]
dp[2] = max(dp[1], dp[0] + wine[2], wine[1] + wine[2])

for i in range(3, n):
    #max{자신을 포함x, 자신을 포함하는데 이전꺼x, 자신 포함하는데 이전꺼 o)
    dp[i] = max(dp[i-1], dp[i-2] + wine[i], dp[i-3] + wine[i-1] + wine[i])

print(dp[n-1])

File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

//DP
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] wine = new int[n];
        for (int i = 0; i < n; i++) {
            wine[i] = Integer.parseInt(br.readLine());
        }

        int[] dp = new int[n];
        dp[0] = wine[0];
        if (n >= 2) {
            dp[1] = wine[0] + wine[1];
        }
        if (n >= 3) {
            dp[2] = Math.max(dp[1], Math.max(wine[0] + wine[2], wine[1] + wine[2]));
        }
        for (int i = 2; i < n; i++) {
            dp[i] = Math.max(dp[i], dp[i - 1]);  //i번째 포도주를 안 마신 경우

            //i번째 포도주 마시고, i-1번째 포도주는 안 마신 경우
            dp[i] = Math.max(dp[i], wine[i] + dp[i - 2]);

            //i번째 포도주 마시고, i-1번째 포도주도 마시는 경우(i-2번째 포도주를 안마셔야함)
            if (i >= 3) {
                dp[i] = Math.max(dp[i], wine[i] + wine[i - 1] + dp[i - 3]);
            }
        }
        System.out.println(dp[n - 1]);
    }
}
